bitcoin/src/crypto/key.rs

```diff
@@ -863,61 +890,84 @@ impl From<TweakedKeypair> for TweakedPublicKey {
     #[inline]
     fn from(pair: TweakedKeypair) -> Self { TweakedPublicKey::from_keypair(pair) }
 }
-/// A key-related error.
+/// Everyting perfect from here
+/// Error generated from WIF key format.
 #[derive(Debug, Clone, PartialEq, Eq)]
 #[non_exhaustive]
-pub enum Error {
+pub enum FromWifError {
     /// A base58 error.
     Base58(base58::Error),
-    /// A secp256k1 error.
-    Secp256k1(secp256k1::Error),
+    /// A secp256k1 error
+    Secp256k1(secp256k1::Error)
+}
+
+/// Error returned while generating key from slice.
+#[derive(Debug, Clone, PartialEq, Eq)]
+#[non_exhaustive]
+pub enum FromSliceError {
+    /// A base 58 error.
+    Base58(base58::Error),
```

Comment:
Why is this variant here? `FromSlice` implies not from string, which implies not WIF, which implies this error variant should never be returned. If it is, it's an error.

Reply:
Okay I get it now. `FromSliceError` should be returning errors related to `Slices` only. But how should we propogate errors from the function [here](https://github.com/rust-bitcoin/rust-bitcoin/blob/6702f1a144801dc5e0c7779297311e76e4672933/bitcoin/src/crypto/key.rs#L174). One of the solution is to implement a new `ErrorType` that can contain both the `FromSliceError` and `FromWIFError` like this :

```rust
pub enum Error {
    SliceError(FromSliceError),
    WifError(FromWifError),
}
```
Or is there any other implementation that we can do? Please let me know.

Reply:
That function should return `FromSliceError` since it's converting from slice. `FromSliceError` must have `InvalidLength`, `InvalidKeyPrefix`, and `Secp256k1` variants.

Reply:
But then how do we check for the following `base58` error?
```rust
    let compressed = match data.len() {
        33 => true,
        65 => false,
        len => {
            return Err(base58::Error::InvalidLength(len).into());
        }
    };
```
One solution could be to return `InvalidLength` here also. Sorry if these are some basic questions.

Reply:
Yes, do that.